### DISCUSSION TITLE
Is it OK to make custom meshes to use on the porting process of the maps to 1.9.9?

### DISCUSSION THREAD

**Discussion by SniperGoth** [redeclipse/base#1123] [Help and Support] [answered]
Something i was wondering...
What if someone decides to make let's say, a model to replace a bridge (kargo, depot and others, im looking at you), in the process of porting maps from 1.6.0 to 1.9.9, is it a open option for us level designers/3d artists, or do we need to stick with octatree geometry?

**Comment by graphitemaster** [comment]
Why would you make a model to replace something if you could just make the contents in the engine with cubes? Is it a placement thing? We've had prefabs for a long time so you can prototype this stuff outside a map perfectly fine. I don't understand the question.

**Comment by SniperGoth** [comment]
Im a 3d artist that just wants to start using meshes for making levels, like on other engines (make things first with BSP/octatree, then replace with static meshes), since we have the tools to do that now (proper collision, rendering etc), so that's why im gonna start using way more models for that purpose.
That's why i asked this question in particular, as using only cubes just isn't doing it anymore to me.

**Comment by IceflowRE** [comment]
If it works and the engine gives enough fps why not?

**Comment by qreeves** [comment, marked as answer]
The question is not whether you *can* but whether you *should*. There are actually multiple things to consider here:

* While better support for models is added, it always comes at a performance cost. Cubes will always be much faster. That's the point of the engine.
* If your content doesn't have any real value outside the map it is designed for, then you're bloating the package for one map.
* Models are much harder for us to maintain and improve over regular map geometry and standard features everyone is able to accomplish. I can't model so I can't maintain these things for the future.

**Comment by SniperGoth** [comment]
Yeah, i know about the possible issues.
That's why im asking about this, as im going to study the maps that would benefit greatly with this approach, try to make meshes that don't fit exactly in 1 place (thinking about modular ones).
As everyone knows, i do more general stuff, props, in this case.
So this will be interesting to say the least...

**Comment by no-lex** [comment]
Reasons us mappers would want to use mmodels rather than geometry:

- **Cannot texture at angles other than multiples of 90 degrees** (vrotate);

    Means that objects like crates, boxes, etc. have to be oriented parallel to the x/y/z planes even if the geometry is achievable within the engine. Creating custom textures rotated specifically for a scene is wasteful and time consuming and is unlikely to be worth the effort and space it consumes.
If one wanted e.g. to not have the shipping containers in Depot be absolutely parallel to each other, it would require custom textures so that the faces of the crates align correctly to the textures. Likewise, it is a lot faster to get scenes looking nice with mapmodels; moving your crate 0.5m to the left requires voffsetting the textures to match and is quite time consuming.

- **Geometry not possible in Cube; angles other than 90/45/27/18/14/11/9/8/7 degrees are not possible; geometry cannot intersect except at cube grid edges; only one texture can be applied per face, etc**

    These are unavoidable when working with organic materials (bushes, plants, trees) and are problematic when designing levels, as floors have to be a multiple of 2 cubes seperated vertically in order to have ramps that travel straight (Cargo e.g. has kinks in its bridges because there is no way to make them straight in cube geometry). Cube geometry also makes it difficult to impossible to implement compound curvature or other surfaces which are necessary to make a map feel less like a set of right angle corridors and more like a real environment. While it is possible to work around these issues, it becomes very difficult at times to make details in Cube geometry.

    *double texturing is an exploit and has issues

- **Geometry is difficult to make, requires at least an order of magnitude more geometry to make in cube form, or requires exploits.**

    Things like slanted cylinders and slanted windows require several times more geometry and require intimate knowledge of Cube's limitations in order to create, making them inaccessible to all but the most experienced mappers. Additionally, geometry like inside slanted corners and double textured surfaces have problems (like not mipping properly and having texture stretch issues) and is required to do such mysterious things as joining two slanted surfaces cleanly.

I'm not saying these problems necessarily justify the use of mapmodels as standard geometry, or even as a common feature in maps, but I do want to point out that there are valid reasons to use mapmodels instead of trying to work around the Cube engine's issues with certain pieces of geometry. Working solely within the constraints of the octree system can be very frustrating at times, and if we are to put more focus on improving the aesthetics of the game, cube geometry alone will not be adequate.
